pkg/xds/generator/transparent_proxy_generator.go

```diff
@@ -37,16 +39,47 @@ func (_ TransparentProxyGenerator) Generate(ctx xds_context.Context, proxy *mode
 	if err != nil {
 		return nil, errors.Wrapf(err, "could not generate cluster: pass_through")
 	}
+
+	inboundPassThroughClusterName := "inbound:pass_through"
+	inboundPassThroughCluster, err := envoy_clusters.NewClusterBuilder().
+		Configure(envoy_clusters.PassThroughCluster(inboundPassThroughClusterName)).
+		Configure(envoy_clusters.UpstreamBindConfig("127.0.0.6", 0)).
+		Build()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not generate cluster: %s", inboundPassThroughClusterName)
+	}
+
+	inboundListenerName := names.GetInboundListenerName("0.0.0.0", 15006)
+	inboundListener, err := envoy_listeners.NewListenerBuilder().
+		Configure(envoy_listeners.OutboundListener(inboundListenerName, "0.0.0.0", 15006)).
+		Configure(envoy_listeners.FilterChain(envoy_listeners.NewFilterChainBuilder().
+			Configure(envoy_listeners.TcpProxy(inboundPassThroughClusterName, envoy_common.ClusterSubset{ClusterName: inboundPassThroughClusterName})))).
+		Configure(envoy_listeners.OriginalDstForwarder()).
+		Build()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not generate listener: %s", inboundListenerName)
+	}
+
 	return []*model.Resource{
-		&model.Resource{
+		{
 			Name:     "catch_all",
 			Version:  proxy.Dataplane.Meta.GetVersion(),
 			Resource: listener,
 		},
-		&model.Resource{
+		{
 			Name:     "pass_through",
 			Version:  proxy.Dataplane.Meta.GetVersion(),
 			Resource: cluster,
 		},
+		{
+			Name:     inboundListenerName,
+			Version:  proxy.Dataplane.Meta.GetVersion(),
+			Resource: inboundListener,
+		},
+		{
+			Name:     inboundPassThroughClusterName,
+			Version:  proxy.Dataplane.Meta.GetVersion(),
+			Resource: inboundPassThroughCluster,
+		},
```

Comment:
Can you help me understand why we only generate this for inbound? and why do we need old ones? Are old ones used only for outbound now?

Reply:
iptables rules redirect inbound traffic to 15006 and outbound traffic to 15001. That's why we need listeners for both ports. Both listeners send traffic to pass_through clusters, but in case 15006 it also binds connection to 127.0.0.6.